registry/storage/driver/gcs/gcs.go

```diff
@@ -864,33 +818,33 @@ func startSession(client *http.Client, bucket string, name string) (uri string,
 	return uri, err
 }
 
-func putChunk(client *http.Client, sessionURI string, chunk []byte, from int64, totalSize int64) (int64, error) {
-	bytesPut := int64(0)
-	err := retry(func() error {
-		req, err := http.NewRequest(http.MethodPut, sessionURI, bytes.NewReader(chunk))
-		if err != nil {
-			return err
-		}
-		length := int64(len(chunk))
-		to := from + length - 1
-		size := "*"
-		if totalSize >= 0 {
-			size = strconv.FormatInt(totalSize, 10)
-		}
-		req.Header.Set("Content-Type", "application/octet-stream")
-		if from == to+1 {
-			req.Header.Set("Content-Range", fmt.Sprintf("bytes */%v", size))
-		} else {
-			req.Header.Set("Content-Range", fmt.Sprintf("bytes %v-%v/%v", from, to, size))
-		}
-		req.Header.Set("Content-Length", strconv.FormatInt(length, 10))
+func (w *writer) putChunk(ctx context.Context, sessionURI string, chunk []byte, from int64, totalSize int64) (int64, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodPut, sessionURI, bytes.NewReader(chunk))
+	if err != nil {
+		return 0, err
+	}
+	length := int64(len(chunk))
+	to := from + length - 1
+	size := "*"
+	if totalSize >= 0 {
+		size = strconv.FormatInt(totalSize, 10)
+	}
+	req.Header.Set("Content-Type", blobContentType)
+	if from == to+1 {
+		req.Header.Set("Content-Range", fmt.Sprintf("bytes */%v", size))
+	} else {
+		req.Header.Set("Content-Range", fmt.Sprintf("bytes %v-%v/%v", from, to, size))
+	}
+	req.Header.Set("Content-Length", strconv.FormatInt(length, 10))
 
-		resp, err := client.Do(req)
+	bytesPut := int64(0)
+	err = retry(func() error {
+		resp, err := w.driver.client.Do(req)
 		if err != nil {
 			return err
 		}
 		defer resp.Body.Close()
-		if totalSize < 0 && resp.StatusCode == 308 {
+		if totalSize < 0 && resp.StatusCode == http.StatusPermanentRedirect {
```

Comment:
Looks like this was added in https://github.com/distribution/distribution/commit/7162cb19c6c28faf4c5da9671f7839619ba66ea6 (https://github.com/distribution/distribution/pull/1438), but  I actually had to Google to see what's the difference between a `308` and `301`;

> The main difference between the 301 and 308 redirects is that when a 308 redirect code is specified, the client must repeat the exact same request (POST or GET) on the target location. For 301 redirect, the client may not necessarily follow the exact same request.

So the difference here would be that a `308` MUST preserve the request type (i.e. `POST` or `GET`), but "exact same request"  (my interpretation) would also mean "preserve all headers", which _MAY_ include authentication headers, correct?

Wondering if (in a follow-up) we should explicitly call out why a 308 is used here.

Reply:
> So the difference here would be that a 308 MUST preserve the request type (i.e. POST or GET), but "exact same request" (my interpretation) would also mean "preserve all headers", which MAY include authentication headers, correct?

That's my assumption, too.